inc/Config/DataSource/HttpDataSource.php

```diff
@@ -16,94 +14,97 @@
+	protected const SERVICE_NAME = REMOTE_DATA_BLOCKS_GENERIC_HTTP_SERVICE;
+	protected const SERVICE_SCHEMA_VERSION = 1;
 
-	final private function __construct( protected array $config ) {}
-
-	abstract public function get_display_name(): string;
-
-	abstract public function get_endpoint(): string;
+	final public function get_display_name(): string {
+		return $this->config['display_name'];
+	}
 
-	abstract public function get_request_headers(): array|WP_Error;
+	public function get_endpoint(): string {
+		return $this->config['endpoint'];
+	}
 
-	public function get_image_url(): ?string {
-		return null;
+	public function get_request_headers(): array|WP_Error {
+		return $this->get_or_call_from_config( 'request_headers' ) ?? [];
 	}
 
-	/**
-	 * Get the service name.
-	 */
-	public function get_service(): ?string {
-		return $this->config['service'] ?? null;
+	public function get_image_url(): ?string {
+		return $this->config['image_url'] ?? null;
 	}
 
-	public function get_uuid(): string {
-		return $this->config['uuid'];
+	final public function get_service_name(): string {
+		return static::SERVICE_NAME;
 	}
 
 	/**
 	 * @inheritDoc
+	 *
+	 * NOTE: This method uses late static bindings to allow child classes to
+	 * define their own validation schema.
 	 */
-	final public static function get_config_schema(): array {
-		$schema = DataSourceInterface::BASE_SCHEMA;
+	public static function from_array( array $config, ?ValidatorInterface $validator = null ): self|WP_Error {
+		$service_config = $config['service_config'] ?? [];
+		$validator = $validator ?? new Validator( static::get_service_config_schema() );
+		$validated = $validator->validate( $service_config );
 
-		if ( isset( static::SERVICE_SCHEMA['properties'] ) ) {
-			$schema['properties'] = array_merge( DataSourceInterface::BASE_SCHEMA['properties'], static::SERVICE_SCHEMA['properties'] );
+		if ( is_wp_error( $validated ) ) {
+			return $validated;
 		}
 
-		return $schema;
+		return parent::from_array(
+			array_merge(
+				static::map_service_config( $service_config ),
+				[
+					// Store the exact data used to create the instance to preserve determinism.
+					'service' => static::SERVICE_NAME,
+					'service_config' => $service_config,
+					'uuid' => $config['uuid'] ?? null,
+				]
+			)
+		);
 	}
 
 	public static function from_uuid( string $uuid ): DataSourceInterface|WP_Error {
-		$config = DataSourceCrud::get_by_uuid( $uuid );
+		$config = DataSourceCrud::get_config_by_uuid( $uuid );
 
-		if ( ! $config ) {
-			return new WP_Error( 'data_source_not_found', __( 'Data source not found.', 'remote-data-blocks' ), [ 'status' => 404 ] );
+		if ( is_wp_error( $config ) ) {
+			return $config;
 		}
 
 		return static::from_array( $config );
 	}
 
 	/**
 	 * @inheritDoc
-	 * @psalm-suppress ParamNameMismatch reason: we want the clarity provided by the rename here
+	 *
+	 * TODO: Do we need to sanitize this to prevent leaking sensitive data?
```

Comment:
I may be misunderstanding the security concern here, but I don't think so. We don't currently expose this in the UI (I think), we just slurp it into `ConfigStore` or use values in code. If a user can run arbitrary code to dump `::to_array()`, they'd also have the ability to read `define`s or env keys or WP config values anyhow. Am I missing something?

Reply:
> We don't currently expose this in the UI (I think)

@alecgeatches Not visually, but anyone that can access the REST endpoint that supports the plugin settings screen (`GET remote-data-blocks/data-sources`) can view the full service config, including credentials. Currently this is limited to `manage_options` but I imagine there will be a push to lower that substantially.

This is currently intentional-by-design, because there is front-end code that needs those credentials in order to call the APIs to do schema discovery. However, we probably need to redact sensitive credentials and conduct these requests via the REST API, so that we don't expose them.

Reply:
They even make their way into the DOM, since they are used to populate password fields when a user edits an existing data source.